packages/language-server/src/core/frontmatterHolders.ts

```diff
@@ -0,0 +1,177 @@
+import { yaml2ts } from '@astrojs/yaml2ts';
+import {
+	type CodeMapping,
+	type LanguagePlugin,
+	type VirtualCode,
+	forEachEmbeddedCode,
+} from '@volar/language-core';
+import type ts from 'typescript';
+import type { URI } from 'vscode-uri';
+
+export const SUPPORTED_FRONTMATTER_EXTENSIONS = { md: 'markdown', mdx: 'mdx', mdoc: 'mdoc' };
+export const SUPPORTED_FRONTMATTER_EXTENSIONS_KEYS = Object.keys(SUPPORTED_FRONTMATTER_EXTENSIONS);
+const SUPPORTED_FRONTMATTER_EXTENSIONS_VALUES = Object.values(SUPPORTED_FRONTMATTER_EXTENSIONS);
+
+export type CollectionConfig = {
+	folder: URI;
+	config: {
+		collections: {
+			hasSchema: boolean;
+			name: string;
+		}[];
+		entries: Record<string, string>;
+	};
+};
+
+function getCollectionName(collectionConfigs: CollectionConfig[], fileURI: string) {
+	for (const collection of collectionConfigs) {
+		if (collection.config.entries[fileURI]) {
+			return collection.config.entries[fileURI];
+		}
+	}
+}
+
+export function getFrontmatterLanguagePlugin(
+	collectionConfigs: CollectionConfig[],
+): LanguagePlugin<URI, FrontmatterHolder> {
+	return {
+		getLanguageId(scriptId) {
+			const fileType = SUPPORTED_FRONTMATTER_EXTENSIONS_KEYS.find((ext) =>
+				scriptId.path.endsWith(`.${ext}`),
+			);
+
+			if (fileType) {
+				return SUPPORTED_FRONTMATTER_EXTENSIONS[
+					fileType as keyof typeof SUPPORTED_FRONTMATTER_EXTENSIONS
+				];
+			}
+		},
+		createVirtualCode(scriptId, languageId, snapshot) {
+			if (SUPPORTED_FRONTMATTER_EXTENSIONS_VALUES.includes(languageId)) {
+				return new FrontmatterHolder(
+					scriptId.fsPath.replace(/\\/g, '/'),
+					languageId,
+					snapshot,
+					getCollectionName(
+						collectionConfigs,
+						decodeURIComponent(scriptId.toString()).toLowerCase(),
```

Comment:
Volar / TypeScript / who knows who unfortunately does some processing on URI that's not expected / consistent (between platforms). 

Decoding and lowercasing as proved to be the only way to get it working everywhere, quite annoying.

Reply:
You might want to leave a comment in the code that explains this behaviour